### PR TITLE
Bug 1553424 - Fix C example (was: log on invalid handles, but don't crash)

### PR DESCRIPTION
**Pull request by badboy**
We never return an error and we certainly don't expect too many.
Yet we should log them _when we_ encounter them.

In debug mode we crash, in release we don't (that's what `debug_assert` is).
Now for the handle thing, do we want to avoid crashing even in debug mode when we get an invalid one? 
It certainly points to an error which we might want to see in debug (and the thing that made us aware, calling `set_upload_enabled` before init will get fixed in another way anyway).

Plus, so far we initialize logging during init, saving us a check at other points, but this means our `invalid handle used` log might actually get lost when it happens before the init.

wdyt?

### Pull Request checklist ###

- [ ] **Quality**: This PR builds and tests run cleanly
  - `make test` runs without emitting any warnings
  - `make lint` runs without emitting any errors
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry or an explanation of why it does not need one
  - Any breaking changes to language binding APIs are noted explicitly

### CODE REVIEW

File: glean-core/ffi/src/handlemap_ext.rs [hunk 1]
```diff
@@ -111,7 +112,20 @@ impl<T> HandleMapExtension for ConcurrentHandleMap<T> {
...
         let res = self.call_with_output(&mut error, h, callback);
-        debug_assert!(error.get_code().is_success());
+
+        if !error.get_code().is_success() {
+            if let ErrorCode::INVALID_HANDLE = error.get_code() {
+                // Intentionally left blank
+                // ffi-support already logged a warning.
+            } else {
+                log::warn!(
+                    "Encounterend an unexpected error. Code: {:?}",
```
Comment (mdboom) [inline glean-core/ffi/src/handlemap_ext.rs]:
```suggestion
                    "Encountered an unexpected error. Code: {:?}",
```

File: glean-core/ffi/src/handlemap_ext.rs [hunk 2]
```diff
@@ -122,7 +136,20 @@ impl<T> HandleMapExtension for ConcurrentHandleMap<T> {
...
         let res = self.call_with_output_mut(&mut error, h, callback);
-        debug_assert!(error.get_code().is_success());
+
+        if !error.get_code().is_success() {
+            if let ErrorCode::INVALID_HANDLE = error.get_code() {
+                // Intentionally left blank
+                // ffi-support already logged a warning.
+            } else {
+                log::warn!(
+                    "Encounterend an unexpected error. Code: {:?}",
```
Comment (mdboom) [inline glean-core/ffi/src/handlemap_ext.rs]:
```suggestion
                    "Encountered an unexpected error. Code: {:?}",
```

**Comment by badboy** [comment]
> I like the idea of crashing when errors hit the FFI boundary in Debug mode, and just logging in release, so 👍 from me, but definitely let @Dexterp37 chime in before merging.

That's the thing: if we want to keep that we don't need this PR (except fixing the C app)

**Comment by Dexterp37** [comment]
> 
> 
> We never return an error and we certainly don't expect too many.
> Yet we should log them _when we_ encounter them.
> 
> In debug mode we crash, in release we don't (that's what `debug_assert` is).
> Now for the handle thing, do we want to avoid crashing even in debug mode when we get an invalid one?
> It certainly points to an error which we might want to see in debug (and the thing that made us aware, calling `set_upload_enabled` before init will get fixed in another way anyway).

I agree: we should keep crashing, in debug mode, on these kind of problems as they point to specific bigger errors.

> Plus, so far we initialize logging during init, saving us a check at other points, but this means our `invalid handle used` log might actually get lost when it happens before the init.

Good point. Any chance we could have that *always* initialized when the library is loaded, on the Rust side?

@badboy so this keeps the crash in debug and additionally prints a message (for release builds), but doesn't fix the log init, right?

**Comment by badboy** [comment]
> Good point. Any chance we could have that _always_ initialized when the library is loaded, on the Rust side?

Yes we can. [After loading the lib, just as a-s does](Good point. Any chance we could have that always initialized when the library is loaded, on the Rust side?).

> @badboy so this keeps the crash in debug and additionally prints a message (for release builds), but doesn't fix the log init, right?

This PR does not crash anymore on invalid handles (e.g. handle=0 when things get called before we created the glean object), but crashes for other bugs.
Without the PR we crash even for invalid handles (IMO the way to go)

**Comment by Dexterp37** [comment]
> 
> 
> > Good point. Any chance we could have that _always_ initialized when the library is loaded, on the Rust side?
> 
> Yes we can. [After loading the lib, just as a-s does](Good point. Any chance we could have that always initialized when the library is loaded, on the Rust side?).

Cool, let's do this.

> > @badboy so this keeps the crash in debug and additionally prints a message (for release builds), but doesn't fix the log init, right?
> 
> This PR does not crash anymore on invalid handles (e.g. handle=0 when things get called before we created the glean object), but crashes for other bugs.
> Without the PR we crash (and log) even for invalid handles (IMO the way to go)

Yes, I agree. Let's crash and log for invalid handles, no special handling.

**Comment by badboy** [comment]
D'oh! Managed to copy text instead of the link in the comment above. Fixed that.
Will update this PR to only fix the C example and do logging in a follow-up

**Pull request merged**